modules/navier_stokes/include/fv/kernels/NSFVBase.h

```diff
@@ -0,0 +1,108 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MooseConfig.h"
+
+#ifdef MOOSE_GLOBAL_AD_INDEXING
+
+#include "MooseTypes.h" // THREAD_ID
+#include "FVUtils.h"
+
+#include <unordered_map>
+#include <vector>
+
+template <typename>
+class MooseVariableFV;
+class SubProblem;
+
+/**
+ * Common base that both NSFV kernels and boundary conditions can inherit from
+ */
+class NSFVBase
+{
+protected:
+  static InputParameters validParams();
+  NSFVBase(const InputParameters & params);
+
+  /**
+   * pressure variable getter
+   */
+  const MooseVariableFV<Real> * pVar() const { return _p_var; }
+
+  /**
+   * x-component of velocity. May also be radial *or* axial velocity in RZ coordinates
+   */
+  const MooseVariableFV<Real> * uVar() const { return _u_var; }
+
+  /**
+   * y-component of velocity. May also be radial *or* axial velocity in RZ coordinates
+   */
+  const MooseVariableFV<Real> * vVar() const { return _v_var; }
+
+  /**
+   * z-component of velocity
+   */
+  const MooseVariableFV<Real> * wVar() const { return _w_var; }
+
+  /**
+   * The dynamic viscosity
+   */
+  Real mu() const { return _mu; }
+
+  /**
+   * The density
+   */
+  Real rho() const { return _rho; }
+
+  /**
+   * Returns the Rhie-Chow 'a' coefficient for the requested element \p elem
+   */
+  const ADReal & rcCoeff(const Elem & elem) const;
+
+  /**
+   * Clear the RC 'a' coefficient cache
+   */
+  void clearRCCoeffs() { _rc_a_coeffs[_nsfv_tid].clear(); }
+
+private:
+  const SubProblem & _nsfv_subproblem;
+
+  const THREAD_ID _nsfv_tid;
+
+  /// A map from elements to the 'a' coefficients used in the Rhie-Chow interpolation. The size of
+  /// the vector is equal to the number of threads in the simulation
+  static std::vector<std::unordered_map<const Elem *, ADReal>> _rc_a_coeffs;
```

Comment:
It seems like in a multi-app context, this could be problematic even if we put it on `NavierStokesApp` in cases where main and sub applications are both `NavierStokesApps`...but yea if we put it on `NavierStokesApp` then I guess we would no longer make it `static` and then it would be fine... :+1:

Reply:
Well - I thought it was implied - by "on the app" - I meant on the app instance - not in the app namespace.

Reply:
This won't work even for a single-app case when you try to do navier-stokes with anything other than a `NavierStokesApp`, e.g. if you tried to run navier-stokes with `PronghornApp`

Reply:
Maybe you need to add an app pointer to the cache indexing?

Reply:
Haha this will be beautiful container train. But yes I think that's the cleanest solution I've thought/heard-of to-date

Reply:
Yeah - static/global info can be a pain to manage.